packages/web-client/app/utils/web3-strategies/layer-1-connection-manager.ts

```diff
@@ -0,0 +1,313 @@
+import detectEthereumProvider from '@metamask/detect-provider';
+import Web3 from 'web3';
+import { NetworkSymbol } from './types';
+import WalletConnectProvider from '@walletconnect/web3-provider';
+import WalletConnectQRCodeModal from '@walletconnect/qrcode-modal';
+import config from '../../config/environment';
+import CustomStorageWalletConnect, {
+  clearWalletConnectStorage,
+} from '../wc-connector';
+import { Emitter, SimpleEmitter } from '../events';
+import { WalletProviderId } from '../wallet-providers';
+import { action } from '@ember/object';
+
+const GET_PROVIDER_STORAGE_KEY = (chainId: string | number) =>
+  `cardstack-chain-${chainId}-provider`;
+const WALLET_CONNECT_BRIDGE = 'https://safe-walletconnect.gnosis.io/';
+
+interface ConnectionManagerOptions {
+  chainId: number;
+  networkSymbol: NetworkSymbol;
+}
+
+type ConnectionManagerEvents =
+  | 'connected'
+  | 'disconnected'
+  | 'incorrect-chain'
+  | 'error';
+
+const BROADCAST_CHANNEL_MESSAGES = {
+  DISCONNECTED: 'DISCONNECTED',
+} as const;
+
+/**
+ * # ConnectionManager
+ * This class instantiates a web3 provider given a WalletProviderId. It then handles the EIP-1193
+ * events that come from the web3 provider and filters them for events that we want to handle
+ * in the Dapp:
+ *
+ * - connected: A wallet is connected, or the address is changed
+ * - disconnected: A wallet is disconnected
+ * - incorrect-chain: A wallet is connected to the wrong chain
+ *
+ * It handles cross-tab communication and persistence of connection information across refreshes.
+ * This class does not, at the moment, store any state used directly by the UI besides providerId.
+ */
+export abstract class ConnectionManager
+  implements Emitter<ConnectionManagerEvents> {
+  networkSymbol: NetworkSymbol;
+  chainId: number;
+  simpleEmitter: SimpleEmitter;
+  broadcastChannel: BroadcastChannel;
+  protected provider: any;
+  protected connected = false;
```

Comment:
Don't really like adding this here since the Layer1Chain has its own understanding of connected state - having a wallet address, but I wanted to make sure that we didn't end up having multiple onDisconnect callbacks.